{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "409a3f81",
   "metadata": {},
   "source": [
    "# Demo of USM Renyi Entropy in Python\n",
    "## How to compute continuous Renyi entropy of USM using the pyusm library\n",
    "#### By Katherine Wuestney\n",
    "\n",
    "**Contents**  \n",
    "\n",
    "* [Introduction](#Intro)\n",
    "    - [Terminology Introduction](#termintro)\n",
    "* [Section 1](#Section1) Renyi continuous entropy of USM\n",
    "    - [Section 1.1](#Section1.1) Renyi Continuous Entropy\n",
    "    - [Section 1.2](#Section1.2) Parzen Window Kernel Density Estimation\n",
    "    - [Section 1.3](#Section1.3) Equation of *d*-dimensional Renyi Entropy of USM\n",
    "* [Section 2](#Section2) Demo of usm_entropy module\n",
    "    - [Section 2.1](#Section2.1) Simplification of Entropy Equation\n",
    "* [Section 3](#Section3)\n",
    "    - [Section 3.1](#Section3.1) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf422ab4",
   "metadata": {},
   "source": [
    "<a id='Intro'></a>\n",
    "## Introduction\n",
    "\n",
    "Universal Sequence Maps (USM) [1](#1) are a generalized version of the CGR maps first introduced by Jeffrey [2](#2) and they function as generalized order-free Markov transition matrices of symbolic sequences. This property in particular, allows for us to compute a generalized entropy estimate of the sequence based on the density of USM coordinates themselves. For more background on USM see the notebook demonstration ['demo_usm_make' ](#link to github url). Vinga and Almeida  introduced a continuous quadratic Renyi entropy of the USM of DNA sequences in [3](#3) which demonstrated promising discriminatory power. We build off of their work by implementing their algorithms in python and generalizing their entropy measure to *d*-dimensional USM. Proofs of the generalized equations are included in the theoretical background sections of this notebook. In many aspects the usm_entropy module uses the original javascript and matlab toolkit (found at the github repository [usm.github.com](https://github.com/usm/usm.github.com.git)) as a template, however, we make notable deviations from the original algorithms in order to implement the generalized entropy equation and to accept as input the CGR class generated by the usm_make module. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfe7457d",
   "metadata": {},
   "source": [
    "<a id='termintro'></a>\n",
    "### Terminology Introduction - Use this section as a glossary\n",
    "The topics presented in this notebook are highly cross-disciplinary and as such there are a variety of different terms used in the literature to refer to the same basic concepts. To ensure clarity of discussion we will define the following within the context of symbolic sequence analysis:\n",
    "* **Symbol** - a nominal data element which has no numeric magnitude in the general Euclidean sense. Symbols are common data types encountered in linguistics and natural language processing, information theory and cryptography as well as genomics. A symbol is congruous with a single category of a categorical variable. A symbol can be anything, including numbers, but it does not behave like regular numbers do as it is generally a proxy \"symbolizing\" some other construct.\n",
    "\n",
    "* **Sequence** or **Symbolic sequence** - a set of indexed symbols for which the order of the symbols is integral and thus object of analysis. Sequences are commonly encountered in genomics, time series analysis, linguistics and natural language processing, among others. Synonyms of sequence include strings, series, or sometimes vectors. \n",
    "\n",
    "* **Generating function** - a process or phenomenon which produces or is represented by a sequence of symbols.\n",
    "\n",
    "* **Alphabet** - the set of all possible symbols a generating function may produce. For example, if our sequence is a paragraph of a Charles Dickens novel, our generating function could be considered 19th century English typography, and our alphabet would be the 26 letters of the English alphabet plus each punctuation character and special character in use during that time. The term alphabet is congruous to the term \"state space\" from dynamical systems, in that the alphabet functions as the basic state space of a symbolic generating function. There are many different ways the size of an alphabet is refered to in the literature but for this discussion here we will refer to an alphabet's size as its dimension *d*.\n",
    "\n",
    "* **k-gram** - a subsequence of a longer sequence comprising of k sequential symbols. For example, if our sequence is \"ACTGGCA\", \"TG\" would be a k-gram with k=2. In symbolic sequence analysis we are often most interested in the frequency and patterns of k-grams of various lengths. Synonyms for k-gram include L-tuple, subsequence, words, motifs, sub-strings, or vectors. \n",
    "\n",
    "* **Suffix** - the k-gram occurring at the very end of a sequence. For \"ACTGGCA\", its length 3 suffix is the k-gram \"GCA\".\n",
    "\n",
    "* **Prefix** - the k-gram occurring at the very beginning of a sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9196ce6",
   "metadata": {},
   "source": [
    "<a id='Section1'></a>\n",
    "## Section 1 Renyi Continuous Entropy of USM\n",
    "Early on researchers suggested that the complexity of discrete patterns in a sequence could be measured by measuring the density distribution of the CGR and USM plots and Vinga and Almeida took this a step further in [3](#3) by proposing an entropy measure of USM plots that combines the continuous Renyi quadratic entropy with a Parzen kernel density estimate. The method uses a spherical symmetric Gaussian kernel with mean zero and diagonal covariance matrix $\\Sigma = \\sigma^2I_d$ where $I_d$ is the *dxd* identity matrix. The original paper establishing this measure only tests it for alphabets of size 4 so we must first generalize the equations to alphabets of arbitrary size *d*. The equation given by Vinga and Almeida for 4-dimensional USM is:\n",
    "<a id='Eq1.1'></a>\n",
    "$$\n",
    "\\begin{equation}\n",
    "H_2(USM) = -ln \\frac{1}{N^2} \\sum_{i=1}^{N} \\sum_{j=1}^{N} \\frac{1}{16\\pi^2\\sigma^4}\\exp\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)\n",
    "\\tag{3.1}\n",
    "\\end{equation}\n",
    "$$\n",
    "where $d_{ij}$ is the squared Euclidean distance between the USM coordinates $a_i$ and $a_j$ [3](#3).\n",
    "\n",
    "The next sections introduce the Renyi Continuous Entropy formula, Parzen kernel density formula, and the proof of the Renyi Continuous entropy of *d*-dimensional USM. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b9efe68",
   "metadata": {},
   "source": [
    "<a id='Section3.1'></a>\n",
    "### 1.1 Renyi Continuous Entropy\n",
    "Renyi entropy was instroduced as a generalization of Shannon entropy and includes an order parameter $\\alpha$ which in part, determines the weighted contribution of improbable events to the overall entropy measure. The limit of Renyi entropy $\\lim_{\\alpha \\to 1}H_{\\alpha}(X)$ is Shannon's original entropy measure. The formula for Renyi entropy of a continous probability density function $f(x)$ is:\n",
    "$$\n",
    "H_{\\alpha}(X) = \\frac{1}{1-\\alpha}\\ln\\int f^{\\alpha}(x)dx\n",
    "$$\n",
    "\n",
    "Vinga and Almeida utilize the Renyi quadratic ($\\alpha = 2$) continuous entropy formula for their USM entropy measure [3](#3) which is given here:\n",
    "<a id='Eq1.1.1'></a>\n",
    "$$\n",
    "\\begin{align}\n",
    "H_{\\alpha}(X) & = \\frac{1}{1-2}\\ln\\int f^{2}(x)dx \\\\\n",
    "& = -\\ln\\int f^{2}(x)dx\n",
    "\\end{align}\n",
    "\\tag{1.1.1}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df55309a",
   "metadata": {
    "tags": [
     "remove_cell"
    ]
   },
   "source": [
    "It is worth note that the statistic Sample Entropy which is based on the correlation integral by Grassberger and Procaccia functions as an estimate of the quadratic Renyi entropy since the correlation integral is also an estimate of quadratic Renyi entropy."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdb2cfa1",
   "metadata": {},
   "source": [
    "<a id='Section1.2'></a>\n",
    "### 1.2 Parzen Window Kernel Density Estimation\n",
    "Vinga and Almeida employ a Parzen window method to estimate the probability density function fed to the Renyi continuous entropy formula. The Parzen window method takes a linear combination of smooth, continuous, differentiable weighting functions called the kernel function, one centered on each sample point $a_i$ in the observed sample. The output of this linear combination gives the estimated density function $\\hat{f}(x)$ of the distribution at all values of $x$. This is represented by the equation below where $x$ is the *d*-dimensional vector valued random variable in $\\mathbf{R}^d$, $a = \\left\\{a_1, a_2, \\dots, a_N\\right\\}$ is the sample of $x$ of size $N$, and $\\kappa()$ is the kernel weighting function applied to the distance between a value in $x$ and each $a_i$ in the sample $a$:\n",
    "$$\n",
    "\\hat{f}(x;a) = \\frac{1}{N}\\sum_{i=1}{N}\\kappa(x-a_i)\n",
    "\\tag{1.2.1}\n",
    "$$\n",
    "\n",
    "Vinga and Almeida use the spherical symmetric *d*-dimensional Gaussian kernel $g_d$ with a fixed volume defined by the parameter $\\sigma^2$, the variance of the Gaussian distribution. Recall the general *d*-dimensional Gaussian distribution with mean $\\mu$ and covariance matrix $\\Sigma$ is defined by the equation:\n",
    "$$\n",
    "g_d(x;\\mu,\\Sigma) = \\frac{1}{(2\\pi)^{d/2}|\\Sigma|^{1/2}}e^{\\left(-\\frac{1}{2}(x - \\mu)^T\\Sigma^{-1}(x-\\mu)\\right)}\n",
    "\\tag{1.2.2}\n",
    "$$\n",
    "\n",
    "where $x^T$ is the transpose vector of $x$ and $|\\Sigma|$ is the determinant of the covariance matrix. \n",
    "\n",
    "In the case of the a spherical symmetric Gaussian kernel, the covariance matrix $\\Sigma$ is diagonal and is equivalent to $\\sigma^2I_d$ where $I_d$ is the *dxd* identity matrix and $\\sigma^2$ is a fixed arbitrary variance.\n",
    "\n",
    "If we substitute $g_d(x;0, \\sigma^2I_d)$ for $\\kappa()$ in the kernel density estimate equation we get:\n",
    "<a id='Eq1.2.3'></a>\n",
    "$$\n",
    "\\begin{align}\n",
    "\\hat{f}(x;a) & = \\frac{1}{N}\\sum_{i=1}^{N} g_d(x;0, \\sigma^2I_d) \\\\\n",
    "& = \\frac{1}{N}\\sum_{i=1}^{N}\\frac{1}{(2\\pi)^{d/2}|\\sigma^2I_d|^{1/2}}e^{\\left(-\\frac{1}{2}(x - a_i)^T(\\sigma^2I_d)^{-1}(x-a_i)\\right)} \\\\\n",
    "\\end{align}\n",
    "\\tag{1.2.3}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2195aeea",
   "metadata": {},
   "source": [
    "<a id='Section1.3'></a>\n",
    "### 1.3 Equation of *d*-dimensional Renyi Entropy of USM\n",
    "Next we establish the equation for quadratice Renyi entropy of *d*-dimensional USM.\n",
    "\n",
    "We first substitute the kernel density equation with spherical Gaussian kernel ([Eq. 1.2.3](#Eq1.2.3)] for $f(x)$ in the equation for continuous quadratic Renyi entropy ([Eq. 1.1](#Eq1.1.1)):\n",
    "$$\n",
    "\\begin{align}\n",
    "H_2(USM) & = -\\ln\\int\\left(\\frac{1}{N}\\sum_{i=1}^{N}\\frac{1}{(2\\pi)^{d/2}|\\sigma^2I_d|^{1/2}}e^{\\left(-\\frac{1}{2}(x - a_i)^T(\\sigma^2I_d)^{-1}(x-a_i)\\right)}\\right)^2dx \\\\\n",
    "& = -\\ln\\int\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}\\left(\\frac{1}{(2\\pi)^{d/2}|\\sigma^2I_d|^{1/2}}e^{\\left(-\\frac{1}{2}(x - a_i)^T(\\sigma^2I_d)^{-1}(x-a_i)\\right)}\\right)\\times \\\\\n",
    "& \\qquad \\qquad \\qquad \\qquad \\qquad \\ \\left(\\frac{1}{(2\\pi)^{d/2}|\\sigma^2I_d|^{1/2}}e^{\\left(-\\frac{1}{2}(x - a_j)^T(\\sigma^2I_d)^{-1}(x-a_j)\\right)}\\right)dx\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "Using the constant coefficient rule of integration and the sum rule of integration (the integral of the sums of functions equals the sum of their integrals) we can move the integral operator within the summation operators and we now see that we are taking a convolution of the two Gaussians. \n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "H_2(USM)  = -\\ln\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}\\int & \\left(\\frac{1}{(2\\pi)^{d/2}|\\sigma^2I_d|^{1/2}}e^{\\left(-\\frac{1}{2}(x - a_i)^T(\\sigma^2I_d)^{-1}(x-a_i)\\right)}\\right)\\times \\\\\n",
    "& \\ \\ \\ \\left(\\frac{1}{(2\\pi)^{d/2}|\\sigma^2I_d|^{1/2}}e^{\\left(-\\frac{1}{2}(x - a_j)^T(\\sigma^2I_d)^{-1}(x-a_j)\\right)}\\right)dx\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "The convolution of two Gaussians, $g_d(x;\\mu_1, \\Sigma_1)$ and $g_d(x; \\mu_2, \\Sigma_2)$, is another Gaussian (proof available in the supplementary material of [3](#3)) of the form:\n",
    "\n",
    "$$\n",
    "g_d(x; \\mu_1 + \\mu_2, \\Sigma_1 + \\Sigma_2) = \\frac{1}{(2\\pi)^{d/2}|\\Sigma_1 + \\Sigma_2|^{1/2}}e^\\left(-\\frac{1}{2}(\\mu_1 - \\mu_2)^T(\\Sigma_1 + \\Sigma_2)^{-1}(\\mu_1 - \\mu_2)\\right)\n",
    "$$\n",
    "\n",
    "So here we get\n",
    "\n",
    "$$\n",
    "H_2(USM) = -\\ln\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}\\frac{1}{(2\\pi)^{d/2}|\\sigma^2I_d + \\sigma^2I_d|^{1/2}}e^{\\left(-\\frac{1}{2}(a_i - a_j)^T(\\sigma^2I_d + \\sigma^2I_d)^{-1}(a_i - a_j)\\right)}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8c3b915",
   "metadata": {},
   "source": [
    "Because we are using a fixed kernel volume approach, $\\sigma^2$ is held constant across all kernels in the equation and so we can further simplify $\\sigma^2I_d + \\sigma^2I_d$ to $2\\sigma^2I_d$. Then the following terms become:  \n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "|2\\sigma^2I_d|^{1/2} & = (2^d\\sigma^{2d})^{1/2} \\\\\n",
    "& = 2^{d/2}\\sigma^d\n",
    "\\end{align}\n",
    "$$ \n",
    "\n",
    "and \n",
    "\n",
    "$$\n",
    "(2\\sigma^2I_d)^{-1} = \\frac{1}{2\\sigma^2}I_d\n",
    "$$\n",
    "\n",
    "The former due to the fact the determinant of a *dxd* diagonal matrix is equal to the product of the elements of the principal diagonal which in the case of a scalar matrix, (i.e. $cI_d$), is the scalar constant raised to the *d<sup>th</sup>* power, $|cI_d| = c^d$ [4](#4). The latter due to the fact that the inverse of a diagonal matrix is a diagonal matrix with the reciprocal of the elements of the original matrix as the elements on its principal diagonal, which in this case is $\\frac{1}{2\\sigma^2}$.\n",
    "\n",
    "Plugging these two simplified terms into the entropy equation and we can perform further simplifications based on the properties of matrix-vector multiplication involving scalars and the identity matrix [4](#4):\n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "H_2(USM) & = -\\ln\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}\\frac{1}{(2\\pi)^{d/2}2^{d/2}\\sigma^d}e^{\\left(-\\frac{1}{2}(a_i - a_j)^T\\frac{1}{2\\sigma^2}I_d(a_i - a_j)\\right)} \\\\\n",
    "& = -\\ln\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}\\frac{1}{2^{d/2}\\pi^{d/2}2^{d/2}\\sigma^d}e^{\\left(-\\frac{1}{2 \\cdot 2\\sigma^2}(a_i - a_j)^T(a_i - a_j)\\right)} \\\\\n",
    "& = -\\ln\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}\\frac{1}{2^{d}\\pi^{d/2}\\sigma^d}e^{\\left(-\\frac{1}{4\\sigma^2}(a_i - a_j)^T(a_i - a_j)\\right)} \\\\\n",
    "& = -\\ln\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}\\frac{1}{(2\\pi^{1/2}\\sigma)^d}e^{\\left(-\\frac{1}{4\\sigma^2}(a_i - a_j)^T(a_i - a_j)\\right)}\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "It can be proved that, for $u$ and $v$ *d*-vectors, the product of the vector of the form $(u-v)$ with its transpose $(u-v)^T$ is equivalent to the squared Euclidean distance between $u$ and $v$ (proof omitted). Following the notation of [3](#3), let $d_{ij}$ represent the squared Euclidean distance between $a_i$ and $a_j$, then we rewrite the entropy equation as:\n",
    "<a id='Eq1.3.1'></a>\n",
    "$$\n",
    "H_2(USM) = -\\ln\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}\\frac{1}{(2\\pi^{1/2}\\sigma)^d}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}\n",
    "\\tag{1.3.1}\n",
    "$$\n",
    "\n",
    "which we can see is congruous with [Eq. 1.1](#Eq1.1).  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33e0640c",
   "metadata": {
    "tags": [
     "remove_cell"
    ]
   },
   "source": [
    "Page 175 of [4](#4) discusses theorem of determinant of triangular, and thus by extension diagonal, matrices.  \n",
    "\n",
    "Theorem 5 on page 41 of [4](#4) gives properties of matrix-vector multiplication. Scalar multiplication with vectors is commutative, ie we can move $\\frac{1}{2\\sigma^2}$ to the left of the vector. A vector multiplied by the identity matrix of the same order equals the vector $I_dx = x$."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a76f46d",
   "metadata": {},
   "source": [
    "<a id='Section2'></a>\n",
    "# Section 2 Demo of usm_entropy module\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16fecc45",
   "metadata": {},
   "source": [
    "<a id='Section2.1'></a>\n",
    "### 2.1 Simplification of Entropy Equation for Python Implementation\n",
    "\n",
    "In the implementation of [Eq. 1.3.1](#Eq1.3.1) in the function usm_entropy.renyi2usm() we utilize the function pdist() from the module scipy.spatial.distance which calculates the pairwise distance between  \n",
    "Moving the constants outside of the summation and we can then simplify even further due to the fact that all the pairwise distances of the form $d_{ii} = 0$ and therefore $\\sum_{\\substack{i=j \\\\ i,j = 1}}^{N}e^{\\left(-\\frac{1}{4\\sigma^2} \\cdot \\ 0\\right)} = N$ and the fact that for the remainder of the pairwise distances, $d_{ij} = d_{ji}$, and so the sum of the pairwise distances where $i \\neq j$ can be obtained by $2 \\cdot \\sum_{\\substack{i<j \\\\ i,j = 1}}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}$.\n",
    "\n",
    "Finally we obtain the equation for Renyi entropy of a *d*-dimensional USM of *N* sample coordinates where $d_{ij}$ is the squared Euclidean distance between sample USM coordinates $a_i$ and $a_j$:  \n",
    "<a id='Eq2.1.1'></a>\n",
    "$$\n",
    "H_2(USM) = -\\ln\\frac{1}{N^2(2\\pi^{1/2}\\sigma)^d}\\times\\left[N + 2 \\cdot\\sum_{\\substack{i<j \\\\ i,j = 1}}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}\\right]\n",
    "\\tag{2.1.1}\n",
    "$$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37413c1b",
   "metadata": {},
   "source": [
    "<a id='Section2.2'></a>\n",
    "### 2.2 usm_entropy.renyi2usm()\n",
    "Equation [2.1.1](#Eq2.1.1) is implemented by the function renyi2usm() in the usm_entropy.py module. \n",
    "> renyi2usm(cgr_coords, sig2v, refseq=None, Plot=True, filesave=False, deep_copy=True)  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b0ff5910",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pyusm'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-d9825f98e39f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mpathlib\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mPath\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m \u001b[1;32mimport\u001b[0m \u001b[0mpyusm\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      4\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mpyusm\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0musm_entropy\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'pyusm'"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import pyusm\n",
    "from pyusm import usm_entropy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3d2d0232",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'a': array([1, 2]), 'b': array([3, 4]), 'c': array([5, 6])}\n",
      "(array([1, 2]), array([3, 4]), array([5, 6]))\n",
      "(1, 3, 5) (2, 4, 6)\n",
      "col stack [[1 2]\n",
      " [3 4]\n",
      " [5 6]]\n",
      "np.array [[1 2]\n",
      " [3 4]\n",
      " [5 6]] (3, 2)\n",
      "dict_values([array([1, 2]), array([3, 4]), array([5, 6])])\n",
      "[array([1, 2]), array([3, 4]), array([5, 6])]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[1, 2],\n",
       "       [3, 4],\n",
       "       [5, 6]])"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A = ['a', 'b', 'c']\n",
    "V = np.array([[1, 2], [3, 4], [5, 6]])\n",
    "#V = np.array([[1,0,0], [0,1,0], [0,0,1]])\n",
    "coords = dict(zip(A, V))\n",
    "print(coords)\n",
    "a, vrts = zip(*coords.items())\n",
    "print(vrts)\n",
    "x, y = zip(*vrts)\n",
    "print(x, y)\n",
    "print(\"col stack\", np.column_stack((x, y)))\n",
    "c = np.array(vrts)\n",
    "print(\"np.array\", c, c.shape)\n",
    "d = np.identity(3)\n",
    "vrts = coords.values()\n",
    "print(vrts)\n",
    "print(list(c))\n",
    "#ix = [:, :]\n",
    "c[:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "87e577bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "WindowsPath('C:/Users/Wuestney/Documents/GitHub/pyusm/doc/demo_files')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demo_files = Path.cwd()/ \"demo_files\"\n",
    "demo_files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f5b99ac",
   "metadata": {},
   "source": [
    "<a id='Section3'></a>\n",
    "# Section 3 Asymptotic Behavior of Renyi Entropy by Natural Log of Kernel Variance\n",
    "In the original paper, Vinga and Almeida [3](#3) explore the dependence relationship between the entropy measure and the kernel variance, $\\sigma^2$, with the graph of the function $H_2 = H_2(ln\\sigma^2)$ and prove this graph has two linear asymptotes of the forms $H_2^+ = \\lim_{ln\\sigma^2 \\to +\\infty} H_2(ln\\sigma^2) = m ln\\sigma^2 + b$ and $H_2^- = \\lim_{ln\\sigma^2 \\to -\\infty} H_2(ln\\sigma^2) = m' ln\\sigma^2 + b'$. Below are proofs of the equations of these asymptotes for the d-dimensional USM. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76df4af8",
   "metadata": {},
   "source": [
    "<a id='Section3.1'></a>\n",
    "### 3.1 Asymptote for $\\ln\\sigma^2 \\to +\\infty$\n",
    "We solve for the coefficients, $m$ and $b$, of the asymptote for $ln\\sigma^2 \\to +\\infty$ using the equations given by Vinga and Almeida [3](#3) $m = \\lim_{\\sigma^2 \\to +\\infty} \\frac{H_2(\\sigma^2)}{\\ln\\sigma^2}$ and $b = \\lim_{\\sigma^2 \\to +\\infty} (H_2(\\sigma^2) - m \\ln \\sigma^2)$. \n",
    "\n",
    "We begin with $m$ and substitute [Eq. 1.3.1](#Eq1.3.1) for $H_2(\\sigma^2)$ in the formula. We then simplify the equation by distributing the exponent *d* in the numerator, splitting the numerator into a sum of three log terms per the product rule, then distributing the negative term across the summation, and transforming the negative log of the fractions into the log of the reciprocals per the rule of cologarithms. \n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "m & = \\lim_{\\sigma^2 \\to +\\infty} \\frac{-\\ln\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}\\frac{1}{(2\\pi^{1/2}\\sigma)^d}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}}{\\ln\\sigma^2} \\\\\n",
    "& = \\lim_{\\sigma^2 \\to +\\infty} \\frac{-\\ln\\frac{1}{N^22^d\\pi^{d/2}\\sigma^d}\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}}{\\ln\\sigma^2} \\\\\n",
    "& = \\lim_{\\sigma^2 \\to +\\infty} \\frac{-\\ln\\frac{1}{N^22^d\\pi^{d/2}} -\\ln\\frac{1}{\\sigma^d} - \\ln\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}}{\\ln\\sigma^2} \\\\\n",
    "& = \\lim_{\\sigma^2 \\to +\\infty} \\frac{\\ln(N^22^d\\pi^{d/2}) + \\ln\\sigma^d - \\ln\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}}{\\ln\\sigma^2} \\\\\n",
    "\\end{align}\n",
    "$$\n",
    "From there we can split the larger fraction into a sum of three fractions with common denominator $\\ln\\sigma^2$ and by dealing with each individually we get,\n",
    "<a id='Eq3.1.1'></a>\n",
    "$$\n",
    "\\begin{align}\n",
    "m & = \\lim_{\\sigma^2 \\to +\\infty} \\frac{\\ln(N^22^d\\pi^{d/2})}{\\ln\\sigma^2} + \\frac{\\ln\\sigma^d}{\\ln\\sigma^2} - \\frac{\\ln\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}}{\\ln\\sigma^2} \\\\\n",
    "& = \\frac{d\\ln\\sigma}{2\\ln\\sigma} - \\lim_{\\sigma^2 \\to +\\infty} \\frac{\\ln\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}}{\\ln\\sigma^2} \\\\\n",
    "& =  \\frac{d}{2} - \\lim_{\\sigma^2 \\to +\\infty}\\frac{\\ln N^2}{\\ln\\sigma^2} \\\\\n",
    "& =  \\frac{d}{2} \n",
    "\\end{align}\n",
    "\\tag{3.1.1}\n",
    "$$\n",
    "\n",
    "since the limit of the first fraction as $\\sigma^2 \\to +\\infty$ is 0, as the numerator remains constant while the denominator grows infinitely. The second fraction is simplified to $d/2$ per the power rule and becomes a constant independent of $\\sigma^2$. The limit of the final fraction is also 0 since $\\lim_{\\sigma^2 \\to +\\infty} e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)} = 1$ and $\\sum_{i=1}^{N}\\sum_{j=1}^{N}1 = N^2$, making the numerator independent of $\\sigma^2$ while the denominator tends toward infinity.\n",
    "\n",
    "\n",
    "And doing the same for $b$ we get \n",
    "<a id='Eq3.1.2'></a>\n",
    "$$\n",
    "\\begin{align}\n",
    "b &= \\lim_{\\sigma^2 \\to +\\infty} -\\ln\\frac{1}{N^22^d\\pi^{d/2}\\sigma^d}\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)} - \\frac{d}{2}\\ln\\sigma^2 \\\\\n",
    "&= \\lim_{\\sigma^2 \\to +\\infty} -1\\left(\\ln\\frac{1}{N^22^d\\pi^{d/2}\\sigma^d}\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)} + \\ln\\sigma^d\\right)\\\\\n",
    "&= \\lim_{\\sigma^2 \\to +\\infty} -1\\left(\\ln\\frac{\\sigma^d}{N^22^d\\pi^{d/2}\\sigma^d}\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}\\right)\\\\\n",
    "&= \\lim_{\\sigma^2 \\to +\\infty} -1\\left(\\ln\\frac{\\sigma^d}{2^d\\pi^{d/2}\\sigma^d} + \\ln\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}\\right)\\\\\n",
    "&= -\\ln\\frac{1}{2^d\\pi^{d/2}} - \\lim_{\\sigma^2 \\to +\\infty} \\ln\\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)} \\\\\n",
    "&= \\ln2^d\\pi^{d/2} - \\ln\\frac{N^2}{N^2} \\\\\n",
    "&= \\ln(2\\pi^{1/2})^d \\\\\n",
    "\\end{align}\n",
    "\\tag{3.1.2}\n",
    "$$\n",
    "\n",
    "since the $\\lim_{\\sigma^2 \\to +\\infty} e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)} = 1$, $\\sum_{i=1}^{N}\\sum_{j=1}^{N}1 = N^2$, and $\\ln(1)=0$.\n",
    "\n",
    "Therefore we determine the equation of the linear asymptote of $H_2 =  H_2(ln\\sigma^2)$ as $\\ln\\sigma^2 \\to +\\infty$ as\n",
    "<a id='Eq3.1.3'></a>\n",
    "$$\n",
    "\\begin{align}\n",
    "H_2^+ & = \\frac{d}{2}\\ln\\sigma^2 + \\ln(2\\pi^{\\frac{1}{2}})^d \\\\\n",
    "& = d\\ln\\sigma + d\\ln2\\pi^{\\frac{1}{2}} \\\\\n",
    "& = d\\left(\\ln\\sigma + \\ln2\\pi^{\\frac{1}{2}}\\right) \\\\\n",
    "& = d\\ln\\sigma2\\pi^{\\frac{1}{2}} \\\\\n",
    "& = \\ln\\left(2\\sigma\\pi^{\\frac{1}{2}}\\right)^d \\\\\n",
    "\\end{align}\n",
    "\\tag{3.1.3}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1b429c0",
   "metadata": {
    "tags": [
     "remove_cell"
    ]
   },
   "source": [
    "For b, the first simplification comes from factoring out -1 from the sum, the second simplification comes from the product rule of logarithms we multiply $\\sigma^2$ and $\\ln\\frac{1}{N^22^d\\pi^{d/2}\\sigma^d}\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}$. The third simplification come again from applying the product rule but this time in splitting into a sum of two different log terms (ie we treat $\\frac{\\sigma^d}{N^22^d\\pi^{d/2}\\sigma^d}\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}$ as the product $\\frac{\\sigma^d}{2^d\\pi^{d/2}\\sigma^d} \\times \\frac{1}{N^2}\\sum_{i=1}^{N}\\sum_{j=1}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}$). The fourth simplification comes from applying the rule of cologarithms and the given above."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60518a4e",
   "metadata": {},
   "source": [
    "<a id='Section3.2'></a>\n",
    "### 3.2 Asymptote for $\\ln\\sigma^2 \\to 0^+$\n",
    "Because $H_2 = H_2(ln\\sigma^2)$ is a logarithmic equation it is undefined for $\\sigma^2 < 0$. Therefore we do not have to worry about the $\\lim_{\\sigma^2 \\to -\\infty}$ of the equation but only need to solve for $\\lim_{\\sigma^2 \\to 0^+}$.  \n",
    "\n",
    "We solve for the coefficients, $m'$ and $b'$, of the asymptote equation $H_2^- = \\lim_{ln\\sigma^2 \\to 0^+} H_2(ln\\sigma^2) = m' ln\\sigma^2 + b'$ using the equations given by Vinga and Almeida [3](#3) $m' = \\lim_{\\sigma^2 \\to 0^+} \\frac{H_2(\\sigma^2)}{\\ln\\sigma^2}$ and $b' = \\lim_{\\sigma^2 \\to 0^+} (H_2(\\sigma^2) - m' \\ln \\sigma^2)$.  \n",
    "\n",
    "For $m'$ we begin the same way as for $m$, except using the simplified entropy equation [Eq. 2.1.1](#Eq2.1.1) an alternate form of [Eq. 1.3.1](#Eq1.3.1) which we derived in [Section 2.1](#Section2.1). We then apply the same simplifications applied in the derivation of [Eq. 3.1.1](#Eq3.1.1) and the fact that $\\lim_{x \\to 0^+}\\ln(x) = -\\infty$ and proceed, \n",
    "<a id='Eq3.2.1'></a>\n",
    "$$\n",
    "\\begin{align}\n",
    "m' & = \\lim_{\\sigma^2 \\to 0^+} \\frac{-\\ln\\frac{1}{N^2(2\\pi^{1/2}\\sigma)^d}\\times\\left[N + 2 \\cdot\\sum_{\\substack{i<j \\\\ i,j = 1}}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}\\right]}{\\ln\\sigma^2} \\\\\n",
    "& = \\frac{d}{2} + \\lim_{\\sigma^2 \\to 0^+} \\frac{\\ln(N^22^d\\pi^{d/2})}{\\ln\\sigma^2} - \\frac{\\left[N + 2 \\cdot\\sum_{\\substack{i<j \\\\ i,j = 1}}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}\\right]}{\\ln\\sigma^2} \\\\\n",
    "& = \\frac{d}{2} - \\lim_{\\sigma^2 \\to 0^+} \\frac{N}{\\ln\\sigma^2} \\\\\n",
    "& = \\frac{d}{2} \\\\\n",
    "\\end{align}\n",
    "\\tag{3.2.1}\n",
    "$$\n",
    "\n",
    "since $\\frac{\\ln(N^22^d\\pi^{d/2})}{\\ln\\sigma^2}$ will tend toward $-0$ as $\\sigma^2 \\to 0^+$ and  $\\lim_{\\sigma^2 \\to 0^+} \\sum_{\\substack{i<j \\\\ i,j = 1}}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)} = 0$ since $\\lim_{\\sigma^2 \\to 0^+} -\\frac{1}{4\\sigma^2}d_{ij} = -\\infty$.\n",
    "\n",
    "For $b'$ we follow the same procedure using [Eq. 2.1.1](#Eq2.1.1) and the simplifications applied for [Eq. 3.1.2](#Eq3.1.2):\n",
    "<a id='Eq3.2.2'></a>\n",
    "$$\n",
    "\\begin{align}\n",
    "b' & = \\lim_{\\sigma^2 \\to 0^+} -\\ln\\frac{1}{N^2\\left(2\\pi^{1/2}\\sigma\\right)^d}\\times\\left[N + 2 \\cdot\\sum_{\\substack{i<j \\\\ i,j = 1}}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}\\right] - \\frac{d}{2}\\ln\\sigma^2 \\\\\n",
    "& = -\\ln\\frac{1}{2^d\\pi^{d/2}} - \\lim_{\\sigma^2 \\to 0^+} \\ln\\frac{1}{N^2}\\left[N + 2 \\cdot\\sum_{\\substack{i<j \\\\ i,j = 1}}^{N}e^{\\left(-\\frac{1}{4\\sigma^2}d_{ij}\\right)}\\right] \\\\\n",
    "& = -\\ln\\frac{1}{2^d\\pi^{d/2}} - \\ln\\frac{N}{N^2} \\\\\n",
    "& = \\ln\\left(2\\pi^{1/2}\\right)^d + \\ln\\frac{N^2}{N} \\\\\n",
    "& = d\\ln2\\pi^{1/2} + \\ln N \\\\\n",
    "\\end{align}\n",
    "\\tag{3.2.2}\n",
    "$$\n",
    "\n",
    "Therefore we determine the equation of the linear asymptote of $H_2 =  H_2(ln\\sigma^2)$ as $\\ln\\sigma^2 \\to 0^+$ as\n",
    "<a id='Eq3.2.3'></a>\n",
    "$$\n",
    "\\begin{align}\n",
    "H_2^- & = \\frac{d}{2}\\ln\\sigma^2 + d\\ln2\\pi^{1/2} + \\ln N \\\\\n",
    "& = d\\ln\\sigma + d\\ln2\\pi^{1/2} + \\ln N \\\\\n",
    "& = d\\left(\\ln\\sigma + \\ln2\\pi^{1/2}\\right) + \\ln N \\\\\n",
    "& = d\\left(\\ln\\sigma2\\pi^{1/2}\\right) + \\ln N \\\\\n",
    "& = \\ln\\left(\\sigma2\\pi^{1/2}\\right)^d + \\ln N \\\\\n",
    "& = \\ln N\\left(2\\sigma\\pi^{1/2}\\right)^d \\\\\n",
    "\\end{align}\n",
    "\\tag{3.2.3}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e374e9b5",
   "metadata": {
    "tags": [
     "remove_cell"
    ]
   },
   "source": [
    "For $b'$ the simplification at step 3 is due to the rule of cologs where a negative log equals the log of the reciprocol of the argument."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0268665c",
   "metadata": {},
   "source": [
    "We can check the validity of these asymptote equations by substituting $d = 4$ we obtain the asymptote equations derived in [3](#3).\n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "H_2^+ & = \\ln\\left(2\\sigma\\pi^{\\frac{1}{2}}\\right)^4  \\\\\n",
    "& = \\ln2^4\\sigma^4\\left(\\pi^{\\frac{1}{2}}\\right)^4  \\\\\n",
    "2\\ln\\sigma^2 + \\ln16\\pi^2 & = \\ln16\\sigma^4\\pi^{2}  \\\\\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "H_2^- & = \\ln N\\left(2\\sigma\\pi^{1/2}\\right)^4 \\\\\n",
    "2\\ln\\sigma^2 + \\ln16\\pi^2 + \\ln N & = \\ln 16N\\sigma^4\\pi^{2} \\\\\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "Where the left side of the equations are the corresponding graph asyptote equations in [3](#3)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "eced8080",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2000"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import csv\n",
    "matlab = []\n",
    "py = []\n",
    "with open(demo_files/'cgr_MC0_pymatcompare-2020-06-02.csv', 'r', encoding='utf-8-sig') as fhand:\n",
    "    reader = csv.DictReader(fhand, dialect='excel')\n",
    "    for row in reader:\n",
    "        matlab.append([row['Column1'], row['Column2'], row['Column3'], row['Column4']])\n",
    "        py.append([row['Column5'], row['Column6'], row['Column7'], row['Column8']])\n",
    "with open(demo_files/'og_MC0_coords.csv', 'w') as fwrite:\n",
    "    writer = csv.writer(fwrite)\n",
    "    writer.writerows(matlab)\n",
    "with open(demo_files/'pyusm_MC0_coords_2020-06-02.csv', 'w') as fwrite:\n",
    "    writer = csv.writer(fwrite)\n",
    "    writer.writerows(matlab)\n",
    "    \n",
    "mc0 = np.genfromtxt(demo_files/'og_MC0_coords.csv', delimiter=',')\n",
    "mc0py = np.genfromtxt(demo_files/'pyusm_MC0_coords_2020-06-02.csv', delimiter=',')\n",
    "len(mc0py)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8620ded2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"[('R1.txt', 'renyi2usm_AA.rn2'), ('rand.txt', 'renyi2usm_alea1c.rn2'), ('m5.txt', 'renyi2usm_alea2c.rn2'), ('m7e.txt', 'renyi2usm_alea3c.rn2'), ('m4.txt', 'renyi2usm_ATCG20.rn2'), ('R5.txt', 'renyi2usm_ATCGA.rn2'), ('m3.txt', 'renyi2usm_ATCx.rn2'), ('Es.txt', 'renyi2usm_Es.seq.rn2'), ('MC0.txt', 'renyi2usm_M0.rn2'), ('MC1.txt', 'renyi2usm_MC91.seq.rn2')]\""
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs = ['R1', 'rand', 'm5', 'm7e', 'm4', 'R5', 'm3', 'Es', 'MC0', 'MC1']\n",
    "seqfiles = [seq + '.txt' for seq in seqs]\n",
    "paths = list(Path(r'C:\\Users\\Wuestney\\Documents\\GitHub\\pyusm\\pyusm\\tests\\expected_output').glob('*.rn2'))\n",
    "renyifiles = [path.name for path in paths]\n",
    "filerelat = list(zip(seqfiles, renyifiles))\n",
    "repr(filerelat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "40b13f7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['1.000000e-10', '-4.09740709268997776e+01'],\n",
       " ['1.778279e-10', '-3.98231989820569652e+01'],\n",
       " ['3.162278e-10', '-3.86723220182710676e+01'],\n",
       " ['5.623413e-10', '-3.75214499713380505e+01'],\n",
       " ['1.000000e-09', '-3.63705741437514547e+01'],\n",
       " ['1.778279e-09', '-3.52196995873220473e+01'],\n",
       " ['3.162278e-09', '-3.40688262956174412e+01'],\n",
       " ['5.623413e-09', '-3.29179492317678211e+01'],\n",
       " ['1.000000e-08', '-3.17670770786117984e+01'],\n",
       " ['1.778279e-08', '-3.06161999112846566e+01'],\n",
       " ['3.162278e-08', '-2.94653261463822886e+01'],\n",
       " ['5.623413e-08', '-2.83144512064695313e+01'],\n",
       " ['1.000000e-07', '-2.71635745661599515e+01'],\n",
       " ['1.778279e-07', '-2.60127017076233003e+01'],\n",
       " ['3.162278e-07', '-2.48618237114405609e+01'],\n",
       " ['5.623413e-07', '-2.37109504608230530e+01'],\n",
       " ['1.000000e-06', '-2.25600738448022184e+01'],\n",
       " ['1.778279e-06', '-2.14091979577933138e+01'],\n",
       " ['3.162278e-06', '-2.02583238564345827e+01'],\n",
       " ['5.623413e-06', '-1.91074456291656354e+01'],\n",
       " ['1.000000e-05', '-1.79565724025536824e+01'],\n",
       " ['1.778279e-05', '-1.68056943615971370e+01'],\n",
       " ['3.162278e-05', '-1.56548192904731582e+01'],\n",
       " ['5.623413e-05', '-1.45039435918118791e+01'],\n",
       " ['1.000000e-04', '-1.33530656840902484e+01'],\n",
       " ['1.778279e-04', '-1.22021918887957490e+01'],\n",
       " ['3.162278e-04', '-1.10513128959152347e+01'],\n",
       " ['5.623413e-04', '-9.90043842254380024e+00'],\n",
       " ['1.000000e-03', '-8.74956103471745550e+00'],\n",
       " ['1.778279e-03', '-7.59868382705209733e+00'],\n",
       " ['3.162278e-03', '-6.44780890177324117e+00'],\n",
       " ['5.623413e-03', '-5.29692954507133251e+00'],\n",
       " ['1.000000e-02', '-4.14605521901560081e+00'],\n",
       " ['1.778279e-02', '-2.99517631795371297e+00'],\n",
       " ['3.162278e-02', '-1.84429950619040062e+00'],\n",
       " ['5.623413e-02', '-6.93411534356468628e-01'],\n",
       " ['1.000000e-01', '4.57538249168731426e-01'],\n",
       " ['1.778279e-01', '1.60857983085374223e+00'],\n",
       " ['3.162278e-01', '2.75970700559035098e+00'],\n",
       " ['5.623413e-01', '3.91089757457050125e+00'],\n",
       " ['1', '5.06212966657647456e+00'],\n",
       " ['1.778279e+00', '6.21338718262572609e+00'],\n",
       " ['3.162278e+00', '7.36465969339561610e+00'],\n",
       " ['5.623413e+00', '8.51594086477294887e+00'],\n",
       " ['10', '9.66722698003950676e+00'],\n",
       " ['1.778279e+01', '1.08185158990140682e+01'],\n",
       " ['3.162278e+01', '1.19698064021244015e+01'],\n",
       " ['5.623413e+01', '1.31210977984380914e+01'],\n",
       " ['100', '1.42723896977905245e+01']]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = paths[0]\n",
    "def parse_rn2_file(rn2file):\n",
    "    with open(rn2file, 'r') as fhand:\n",
    "        lines = fhand.readlines()\n",
    "    rn2array = []\n",
    "    row = 0\n",
    "    for line in lines:\n",
    "        if row < 2:\n",
    "            row += 1\n",
    "            continue\n",
    "        else:\n",
    "            row += 1\n",
    "            line = line.rstrip().split('\\t')\n",
    "            #print(line)\n",
    "            rn2array.append([line[0].strip(), line[1].strip()])\n",
    "    return rn2array\n",
    "rn2 = parse_rn2_file(f)\n",
    "rn2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ede7474c",
   "metadata": {},
   "source": [
    "## References\n",
    "<a id='1'></a>\n",
    "<div class=\"csl-entry\">[1] Jeffrey, H. J. (1990). Chaos game representation of gene structure. <i>Nucleic Acids Research</i>, <i>18</i>(8), 2163–2170. https://doi.org/10.1093/nar/18.8.2163</div>\n",
    "<a id='2'></a>\n",
    "<div class=\"csl-entry\">[2] Almeida, J. S., &#38; Vinga, S. (2002). Universal sequence map (USM) of arbitrary discrete sequences. <i>BMC Bioinformatics</i>, <i>3</i>. https://doi.org/10.1186/1471-2105-3-6</div>\n",
    "<a id='3'></a>\n",
    "<div class=\"csl-entry\">[3] Vinga, S., &#38; Almeida, J. S. (2004). Rényi continuous entropy of DNA sequences. <i>Journal of Theoretical Biology</i>, <i>231</i>(3), 377–388. https://doi.org/10.1016/j.jtbi.2004.06.030</div>\n",
    "<a id='4'></a>\n",
    "<div class=\"csl-entry\">[4] David C. Lay, Steven R. Lay, &#38; Judi J. McDonald. (2021). <i>Linear Algebra and Its Applications</i> (6th Edition). Pearson Education, Inc.</div>\n",
    "\n",
    "\n",
    "<a id='7'></a>\n",
    "</div><div class=\"csl-entry\">[7] Almeida, J. S., Carrico, J. A., Maretzek, A., Noble, P. A., &#38; Fletcher, M. (2001). Analysis of genomic sequences by Chaos Game Representation. <i>Bioinformatics</i>, <i>17</i>(5), 429–437.</div>\n",
    "<a id='8'></a>\n",
    "<div class=\"csl-entry\">[8] Almeida, J. S., Carrico, J. A., Maretzek, A., Noble, P. A., &#38; Fletcher, M. (2001). Analysis of genomic sequences by Chaos Game Representation. <i>Bioinformatics</i>, <i>17</i>(5), 429–437.</div>\n",
    "<a id='9'></a>\n",
    "<div class=\"csl-entry\">[9] Almeida, J. S., Grüneberg, A., Maass, W., &#38; Vinga, S. (2012). Fractal MapReduce decomposition of sequence alignment. <i>Algorithms for Molecular Biology</i>, <i>7</i>(1), 12. https://doi.org/10.1186/1748-7188-7-12</div>\n",
    "<a id='10'></a>\n",
    "<div class=\"csl-entry\">[10] Almeida, J. S., &#38; Vinga, S. (2006). Computing distribution of scale independent motifs in biological sequences. <i>Algorithms for Molecular Biology</i>, <i>1</i>(1), 1-11. https://almob.biomedcentral.com/articles/10.1186/1748-7188-1-18</div>\n"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
